pkg/cloudprovider/providers/qingcloud/qingcloud_storage.go

```diff
@@ -0,0 +1,166 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package qingcloud
+
+// See https://docs.qingcloud.com/api/volume/index.html
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/magicshui/qingcloud-go"
+	"github.com/magicshui/qingcloud-go/volume"
+	"k8s.io/kubernetes/pkg/types"
+)
+
+// DefaultMaxQingCloudVolumes is the limit for volumes attached to an instance.
+// TODO: No clear description from qingcloud document
+//const DefaultMaxQingCloudVolumes = 6
+
+// VolumeOptions specifies capacity and type for a volume.
+// See https://docs.qingcloud.com/api/volume/create_volumes.html
+type VolumeOptions struct {
+	CapacityGB int // minimum 10GiB, maximum 500GiB, must be a multiple of 10x
+	VolumeType int // only can be 0, 1, 2, 3
+}
+
+// Volumes is an interface for managing cloud-provisioned volumes
+type Volumes interface {
+	// Attach the disk to the specified instance
+	// Returns the device (e.g. /dev/sdb) where we attached the volume
+	// It checks if volume is already attached to node and succeeds in that case.
+	AttachVolume(volumeID string, nodeName types.NodeName) (string, error)
+
+	// Detach the disk from the specified instance
+	DetachVolume(volumeID string, nodeName types.NodeName) error
+
+	// Create a volume with the specified options
+	CreateVolume(volumeOptions *VolumeOptions) (volumeID string, err error)
+
+	// Delete the specified volume
+	// Returns true if the volume was deleted
+	// If the was not found, returns (false, nil)
+	DeleteVolume(volumeID string) (bool, error)
+
+	// Check if the volume is already attached to the instance
+	VolumeIsAttached(volumeID string, nodeName types.NodeName) (bool, error)
+}
+
+// AttachVolumes implements Volumes.AttachVolume
+func (qc *Qingcloud) AttachVolume(volumeID string, nodeName types.NodeName) (string, error) {
+	attached, err := qc.VolumeIsAttached(volumeID, nodeName)
+	if err != nil {
+		return "", err
+	}
+
+	volumesN := qingcloud.NumberedString{}
+	volumesN.Add(volumeID)
+
+	if !attached {
+		instance := qingcloud.String{}
+		instance.Set(nodeNameToInstanceId(nodeName))
+		_, err := qc.volumeClient.AttachVolumes(volume.AttachVolumesRequest{
+			VolumesN: volumesN,
+			Instance: instance,
+		})
+		if err != nil {
+			return "", err
+		}
+	}
+
+	resp, err := qc.volumeClient.DescribeVolumes(volume.DescribeVolumesRequest{
+		VolumesN: volumesN,
+	})
+	if err != nil {
+		return "", err
+	}
+	if len(resp.VolumeSet) == 0 {
+		return "", fmt.Errorf("volume '%v' miss after attach it", volumeID)
+	}
+
+	dev := resp.VolumeSet[0].Instance.Device
```

Comment:
Again, this is where more contextual info is needed. There could be cases where attach is already on the way but fail to finish because of conflict (multi attach for example). If such cases exist, the device should be released to avoid dead lock.